### PR TITLE
backport: Merge bitcoin#20998, 20832, 21077, 20663, 20915

### PR DESCRIPTION
**Pull request by vijaydasmp**
Bitcoin backports

### CODE REVIEW

File: src/key_io.cpp [hunk 1]
```diff
@@ -58,6 +59,9 @@ CTxDestination DecodeDestination(const std::string& str, const CChainParams& par
             std::copy(data.begin() + script_prefix.size(), data.end(), hash.begin());
             return ScriptHash(hash);
         }
+
+        // Set potential error message.
+        error_str = "Invalid prefix for Base58-encoded address";
     }
     return CNoDestination();
```
Comment (UdjinM6) [inline src/key_io.cpp:66]: 20832:
```suggestion

    // Set error message if address can't be interpreted as Base58.
    if (error_str.empty()) error_str = "Invalid address format";

    return CNoDestination();
```

File: test/functional/rpc_invalid_address_message.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        self.skip_if_no_wallet()
+
+    def test_validateaddress(self):
+        node = self.nodes[0]
+
+        # Base58
+        info = node.validateaddress(BASE58_INVALID_PREFIX)
+        assert not info['isvalid']
+        assert_equal(info['error'], 'Invalid prefix for Base58-encoded address')
+
```
Comment (UdjinM6) [inline test/functional/rpc_invalid_address_message.py:35]: 20832:
```suggestion

        # Invalid address format
        info = node.validateaddress(INVALID_ADDRESS)
        assert not info['isvalid']
        assert_equal(info['error'], 'Invalid address format')

```

File: test/functional/rpc_invalid_address_message.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    def test_getaddressinfo(self):
+        node = self.nodes[0]
+
+        assert_raises_rpc_error(-5, "Invalid prefix for Base58-encoded address", node.getaddressinfo, BASE58_INVALID_PREFIX)
+
```
Comment (UdjinM6) [inline test/functional/rpc_invalid_address_message.py:40]: 20832:
```suggestion

        assert_raises_rpc_error(-5, "Invalid address format", node.getaddressinfo, INVALID_ADDRESS)

```

File: src/rpc/blockchain.cpp [hunk 4]
```diff
@@ -214,19 +214,9 @@ UniValue blockheaderToJSON(const CBlockIndex* tip, const CBlockIndex* blockindex
 
 UniValue blockToJSON(const CBlock& block, const CBlockIndex* tip, const CBlockIndex* blockindex, llmq::CChainLocksHandler& clhandler, llmq::CInstantSendManager& isman, bool txDetails)
 {
-    // Serialize passed information without accessing chain state of the active chain!
-    AssertLockNotHeld(cs_main); // For performance reasons
+    UniValue result = blockheaderToJSON(tip, blockindex, clhandler, isman);
```
Comment (UdjinM6) [inline src/rpc/blockchain.cpp:217]: 21114: a part of #5842

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: test/functional/rpc_invalid_address_message.py (hunk 3 above)
Comment (vijaydasmp) [reply test/functional/rpc_invalid_address_message.py:40]: should we delete line 39
`assert_raises_rpc_error(-5, "Invalid prefix for Base58-encoded address", node.getaddressinfo, BASE58_INVALID_PREFIX)
`

File: src/rpc/blockchain.cpp (hunk 4 above)
Comment (vijaydasmp) [reply src/rpc/blockchain.cpp:217]: droppped 21114

**Review by UdjinM6 — CHANGES_REQUESTED** [review]

File: test/functional/rpc_invalid_address_message.py (hunk 3 above)
Comment (vijaydasmp) [reply test/functional/rpc_invalid_address_message.py:44]: Ok

File: test/functional/rpc_invalid_address_message.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        assert not info['isvalid']
+        assert_equal(info['error'], 'Invalid prefix for Base58-encoded address')
+        # Invalid address format
+        info = node.validateaddress(INVALID_ADDRESS)
+        assert not info['isvalid']
+        assert_equal(info['error'], 'Invalid address format')
+
+    def test_getaddressinfo(self):
+        node = self.nodes[0]
+
```
[the remark this answers is not in the document]
Comment (vijaydasmp) [reply test/functional/rpc_invalid_address_message.py:42]: got it

File: test/functional/rpc_invalid_address_message.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+from test_framework.test_framework import BitcoinTestFramework
+
+from test_framework.util import (
+    assert_equal,
+    assert_raises_rpc_error,
+)
+
+
+BASE58_VALID = 'mipcBbFg9gMiCh81Kj8tqqdgoZub1ZJRfn'
+BASE58_INVALID_PREFIX = '17VZNX1SN5NtKa8UQFxwQbFeFc3iqRYhem'
```
Comment (UdjinM6) [inline test/functional/rpc_invalid_address_message.py:16]: 20832: we need to dashify these
```suggestion
BASE58_VALID = 'yjQ5gLvGRtmq1cwc4kePLCrzQ8GVCh9Gaz'
BASE58_INVALID_PREFIX = 'XpG61qAVhdyN7AqVZQsHfJL7AEk4dPVinc'
```

**Review by UdjinM6 — CHANGES_REQUESTED** [review]
missed one more thing

File: test/functional/rpc_invalid_address_message.py (hunk 6 above)
Comment (vijaydasmp) [reply test/functional/rpc_invalid_address_message.py:16]: Sure

**Review by UdjinM6 — APPROVED** [review]
utACK

**Review by PastaPastaPasta — APPROVED** [review]
utACK for merge via merge commit

**Pull request merged**